libraries/AP_HAL_PX4/CAN.cpp

```diff
@@ -0,0 +1,1035 @@
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN1RX0, can1_irq);
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN1RX1, can1_irq);
+    }
+
+    if (can_number >= 1) {
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN2TX, can2_irq);
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN2RX0, can2_irq);
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN2RX1, can2_irq);
+    }
+}
+
+int PX4CANDriver::init(const uint32_t bitrate, const PX4CAN::OperatingMode mode, uint8_t can_number)
+{
+    int res = 0;
+
+    can_number_ = can_number;
+
+    if (AP_BoardConfig::get_can_debug() >= 2) {
+        printf("PX4CANDriver::init Bitrate %lu mode %d\n\r", static_cast<unsigned long>(bitrate), static_cast<int>(mode));
+    }
+
+    static bool initialized_once = false;
+    if (!initialized_once) {
+        initialized_once = true;
+        if (AP_BoardConfig::get_can_debug() >= 2) {
+            printf("PX4CANDriver::init First initialization\n\r");
+        }
+        initOnce(can_number);
+    }
+
+    /*
+     * CAN1
+     */
+    if (can_number >= 1) {
+        if (AP_BoardConfig::get_can_debug() >= 2) {
+            printf("PX4CANDriver::init Initing iface 0...\n\r");
+        }
+        ifaces[0] = &if0_;               // This link must be initialized first,
+        res = if0_.init(bitrate, mode); // otherwise an IRQ may fire while the interface is not linked yet;
+        if (res < 0) {             // a typical race condition.
+            if (AP_BoardConfig::get_can_debug() >= 2) {
+                printf("PX4CANDriver::init Iface 0 init failed %i\n\r", res);
+            }
+            ifaces[0] = UAVCAN_NULLPTR;
+        }
+    }
+
+    /*
+     * CAN2
+     */
+    if (can_number >= 2) {
+#if UAVCAN_STM32_NUM_IFACES > 1
+        if (AP_BoardConfig::get_can_debug() >= 2) {
+            printf("PX4CANDriver::init Initing iface 1...\n\r");
+        }
+        ifaces[1] = &if1_;                          // Same thing here.
+        res = if1_.init(bitrate, mode);
+        if (res < 0) {
+            if (AP_BoardConfig::get_can_debug() >= 2) {
+                printf("PX4CANDriver::init Iface 1 init failed %i\n\r", res);
+            }
+            ifaces[1] = UAVCAN_NULLPTR;
+        }
+#endif
+    }
+
+    if (AP_BoardConfig::get_can_debug() >= 2) {
+        printf("PX4CANDriver::init CAN drv init OK, res = %d\n\r", res);
+    }
+
+    return res;
+}
+
+PX4CAN* PX4CANDriver::getIface(uint8_t iface_index)
+{
+    if (iface_index < UAVCAN_STM32_NUM_IFACES) {
+        return ifaces[iface_index];
+    }
+    return UAVCAN_NULLPTR;
+}
+
+bool PX4CANDriver::hadActivity()
+{
+    bool ret = if0_.hadActivity();
+#if UAVCAN_STM32_NUM_IFACES > 1
+    ret |= if1_.hadActivity();
+#endif
+    return ret;
+}
+
+bool PX4CANDriver::begin(uint32_t bitrate, uint8_t can_number)
+{
+    if (init(bitrate, PX4CAN::OperatingMode::NormalMode, can_number) >= 0) {
+        initialized_ = true;
+
+        if (p_uavcan != UAVCAN_NULLPTR) {
+            uint16_t UAVCAN_init_tries;
+
+            // TODO: Something
```

Comment:
Are you saying that something needs to be done/changed here but you aren't clear how or you forgot what this todo was about?
If it's the former, then can you just write what needs to be done/changed? Even if you don't know how to do it, a more informative description can give someone an idea.
If it's the latter and you can't remember what it was then we should remove the comment - in the current state it's basically useless.